New: Add visualization table format dropdown

Xuefeng-Zhu opened the pull request:
![image](https://user-images.githubusercontent.com/5875315/39841046-27f4db6e-53a7-11e8-89de-19cb4a5a907c.png)

jkusa commented on packages/visualizations/addon/components/navi-visualizations/table.js:
```diff
@@ -37,7 +37,7 @@ export default Component.extend({
    * @property {Array} classNames - list of component class names
    */
-  classNames: [ 'table-widget' ],
+  classNames: ["table-widget"],
```
We typically use single quotes in JS files. Can we revert the quote changes?

Xuefeng-Zhu replied:
weird. I did not make such changes.

Xuefeng-Zhu replied:
It is caused by a editor plugin I installed recently

jkusa commented on packages/visualizations/addon/components/navi-visualizations/table.js:
```diff
@@ -291,20 +304,22 @@ export default Component.extend({
+    /**
+     * @action updateHeaderFormat
+     */
+    updateHeaderFormat(column, {format}) {
```
Can this be `updateColumnNumberFormat`

jkusa commented on packages/visualizations/addon/components/navi-visualizations/table.js:
```diff
@@ -291,20 +304,22 @@ export default Component.extend({
+     * @action updateHeaderFormat
+     */
+    updateHeaderFormat(column, {format}) {
+      set(column, 'format', format);
```
Can we do something like:
```js
let newColumn = assign({}, column, { format });
this.attrs.onUpdateReport("updateColumn", column)
```

jkusa commented on packages/visualizations/addon/templates/components/navi-visualizations/table.hbs:
```diff
@@ -14,6 +14,15 @@
                 value=(readonly column.displayName)
                 focus-out=(action 'updateHeaderDisplayName' column)
               }}
+              {{#basic-dropdown as |dd| }}
```
Will this be showing up on dateTime and dimension columns?

jkusa commented:
@Xuefeng-Zhu couple general questions/comments:
 - Can we add some padding to the dropdown content?
 - Is the dropdown arrow not aligned with the trigger?
 - Are you adding tests?

Xuefeng-Zhu commented:
> Can we add some padding to the dropdown content?

There is already padding added

> Is the dropdown arrow not aligned with the trigger?

updated

> Are you adding tests?

Will do once finalized this feature

balajimanoharan commented:
General Questions/Comments:
 - Can we decrease the font size to match the either the table font or the header font-size? I feel this font-size is a little too big.
- The screenshot above is format dropdown for `unique Identifiers`? It would be better if we open to the bottom-left so that the dropdown arrow is on the top right of the dropdown. Can we do that please?
- Can we also have the dropdown width match the column width?

jkusa commented:
Can we add more padding to the dropdown?

Xuefeng-Zhu commented:
> Can we also have the dropdown width match the column width?

It seems that the column width is dynamic?

> Do we need the line separators? I don't think they were in the original design.

Do we want to remove it from the number format selector component? Or just hide it in the dropdown?

Xuefeng-Zhu commented:
@balajimanoharan @jkusa Here is the latest screenshot
![image](https://user-images.githubusercontent.com/5875315/39945425-07a58680-5530-11e8-8cae-7dd2a8012587.png)

mattgraham1995 commented on packages/visualizations/addon/templates/components/navi-visualizations/table.hbs:
```diff
@@ -14,6 +14,17 @@
                 value=(readonly column.displayName)
                 focus-out=(action 'updateHeaderDisplayName' column)
               }}
+              {{#if (eq column.type 'metric') }}
```
Do we want the dropdown enabled for `threshold` columns as well?

Xuefeng-Zhu replied:
@jkusa Could you please confirm this?

Xuefeng-Zhu commented:
@balajimanoharan 
This is the latest screenshot

![image](https://user-images.githubusercontent.com/5875315/40008413-dbc7e8e6-5764-11e8-978e-cf0891d52ae6.png)

balajimanoharan commented on packages/visualizations/addon/components/navi-visualizations/table.js:
```diff
@@ -37,7 +38,7 @@ export default Component.extend({
    * @property {Array} classNames - list of component class names
    */
-  classNames: [ 'table-widget' ],
+  classNames: ['table-widget'],
```
Can we reset all these space changes pls?

balajimanoharan commented on packages/visualizations/app/styles/navi-visualizations/visualizations/table.less:
```diff
@@ -219,3 +224,50 @@
   }
 }
+
+.table-header-cell__dropdown-container {
```
Since this is a different component, can we have its own less file?

Xuefeng-Zhu replied:
We do not have an individual component filer for the dropdown, so it is very matching the pattern how we organize the style file.

balajimanoharan replied:
But it is a different component, we don't have a nested style here? This classname is not nested under the table-widget, so it feels like it belongs in its own file.

balajimanoharan commented on packages/visualizations/app/styles/navi-visualizations/visualizations/table.less:
```diff
@@ -219,3 +224,50 @@
+  }
+
+  .number-format-selector__section-header {
+    font-weight: 600;
```
We try to use our bold fonts instead of adding font-weight to properties. Could we do that here too pls?
Maybe `font-family-strong` or `font-family-bold`? https://github.com/yahoo/navi/blob/master/packages/core/app/styles/navi-core/variables/fonts.less#L19

balajimanoharan commented on packages/visualizations/app/styles/navi-visualizations/visualizations/table.less:
```diff
@@ -219,3 +224,50 @@
+
+  .number-format-selector__format-input {
+    font-size: @font-size-base;
+    background-color: @navi-purple-lightest;
```
We have been going with the denali styles nowadays. Could we use a light blue instead like in the mock?

Xuefeng-Zhu commented on packages/visualizations/tests/acceptance/table-test.js:
```diff
@@ -1,5 +1,6 @@
 import { test } from 'qunit';
 import moduleForAcceptance from '../../tests/helpers/module-for-acceptance';
+import { clickTrigger } from 'ember-basic-dropdown/test-support/helpers';
```
We have not upgraded to latest `ember-basic-dropdown ` yet

balajimanoharan replied:
My question was if the app would break if we do use the latest version?

Xuefeng-Zhu replied:
I am not sure, but it may be out of scope of this PR

cythrawll replied:
Where do you see that it's deprecated? I see global function registration is deprecated. which this isn't doing.

Xuefeng-Zhu commented:
@balajimanoharan 

![image](https://user-images.githubusercontent.com/5875315/40079666-3f50275e-584e-11e8-8ce1-12db8c66b405.png)

jkusa commented on packages/visualizations/app/styles/navi-visualizations/visualizations/table.less:
```diff
@@ -219,3 +224,59 @@
+    padding: 0;
+  }
+
+  .number-format-selector__format-list {
```
Should some of these styles go in the base component?

Xuefeng-Zhu replied:
I guess not

jkusa commented on packages/visualizations/app/styles/navi-visualizations/visualizations/table.less:
```diff
@@ -219,3 +224,59 @@
+
+  .number-format-selector__format-input {
+    font-size: @font-size-base;
+    background-color: @brand-100;
```
🔤

Xuefeng-Zhu commented on packages/core/app/styles/navi-core/variables/colors.less:
```diff
@@ -81,3 +81,7 @@
+
+// Brand Palette
+// -----------
+@brand-100: #f5f8fe;
```
from `denali` color spec

balajimanoharan replied:
Can we use a different variable name that says what the color is?
Since this is a type of blue, it should have a name like `navi-blue-{number}`. Or we could just use the blue we already have and add opacity.

balajimanoharan replied:
^ Can we address the comment above pls?

balajimanoharan commented on packages/visualizations/addon/components/number-format-dropdown.js:
```diff
@@ -0,0 +1,21 @@
+import Component from '@ember/component';
```
needs ©️

balajimanoharan commented on packages/visualizations/addon/templates/components/number-format-dropdown.hbs:
```diff
@@ -0,0 +1,9 @@
+{{#basic-dropdown horizontalPosition='right' as |dd| }}
```
©️

Xuefeng-Zhu commented:
@balajimanoharan Sorry for the confusion. Previously, I have not finished addressing comments. They are all addressed now. Please check again. Thank you.

jkusa commented on packages/visualizations/addon/templates/components/number-format-dropdown.hbs:
```diff
@@ -0,0 +1,11 @@
+{{!-- Copyright 2018, Yahoo Holdings Inc. Licensed under the terms of the MIT license. See accompanying LICENSE.md file for
+terms. --}}
+{{#basic-dropdown horizontalPosition='right' onClose=(action 'updateColumnNumberFormat') as |dd| }}
```
Is there a reason why we need to update the format in the column object `onClose`? Can we do this `onUpdateFormat`?

Xuefeng-Zhu replied:
This is caused by the rendering issue we discussed before. If we update the column format immediately, the dropdown will be disappeared.

jkusa replied:
ok

jkusa commented on packages/visualizations/app/styles/navi-visualizations/components/number-format-dropdown.less:
```diff
@@ -0,0 +1,63 @@
+
+    .number-format-selector__doc-link {
+      color: @navi-blue;
+      font-size: @font-size-mid;
```
Can we make this a little smaller?

Xuefeng-Zhu replied:
![image](https://user-images.githubusercontent.com/5875315/40505511-0b0c0572-5f5a-11e8-9d59-5f090ab01f0b.png)

Xuefeng-Zhu replied:
done

jkusa commented on packages/visualizations/tests/acceptance/table-test.js:
```diff
@@ -54,5 +55,21 @@ test('toggle table editing', function (assert) {
+    andThen(function () {
+      assert.ok($('.table-header-cell__dropdown-container').is(':visible'),
+        'Table format dropdown should be visible');
+    });
```
Do we have an end to end test that updates the format and verifies the new format is reflected in the table?

Xuefeng-Zhu replied:
Not yet, since we do not have the table format yet.

jkusa replied:
Can we get the format in then add the test?

Xuefeng-Zhu commented:
@balajimanoharan 

There is the screenshot 
![image](https://user-images.githubusercontent.com/5875315/40505522-102d5d1c-5f5a-11e8-8557-0f3d63a8c2be.png)

jkusa commented:
@Xuefeng-Zhu the latest screenshot looks pretty good, but can we move the drop down a little to the right so the triangle is aligned with the trigger icon?

Xuefeng-Zhu commented:
@jkusa @balajimanoharan 

![image](https://user-images.githubusercontent.com/5875315/40517173-85d84882-5f79-11e8-9b60-a50a1059cefa.png)

jkusa commented on packages/visualizations/app/styles/navi-visualizations/visualizations/table.less:
```diff
@@ -86,6 +86,11 @@
       padding: 5px;
     }
+
+    .number-format-dropdown {
```
`&__number-format-dropdown`

jkusa commented on packages/visualizations/addon/templates/components/navi-visualizations/table.hbs:
```diff
@@ -14,6 +14,9 @@
                 focus-out=(action 'updateHeaderDisplayName' column)
               }}
+              {{#if (or (eq column.type 'metric') (eq column.type 'threshold'))}}
+                {{number-format-dropdown column=column onUpdateReport=onUpdateReport}}
```
`classNames=table-widge__number-format-dropdown`

jkusa replied:
Since this feature isn't ready, do we have a feature flag somewhere to hide this)?

Xuefeng-Zhu replied:
The feature flag has been added in https://github.com/yahoo/navi/pull/91/files#diff-2b01e32ee456dad575a2128ca8e9e300R48

balajimanoharan commented on packages/visualizations/addon/components/number-format-dropdown.js:
```diff
@@ -0,0 +1,43 @@
+   * @property {Array} classNames
+   */
+  classNames: ['number-format-dropdown'],
+
```
Can we add a property 
```
import { readOnly } from '@ember/computed';
format: readOnly('column.format'),
```

that way we won't need the `(or format column.format)` in the hbs. Clarifies things a little.

Xuefeng-Zhu replied:
format is current used as a placeholder for attempting format change, so this will break the implementation

balajimanoharan replied:
How would it break? 
`format` is either `column.format` which the readOnly would solve, or when a format is set by the `number-format-selector`, which sets `format` when something changes.

Xuefeng-Zhu replied:
![image](https://user-images.githubusercontent.com/5875315/40742448-35571a6a-6414-11e8-93b3-8a7484195c59.png)

If we make it readonly,  we will not be able to set it in `onUpdateFormat `

balajimanoharan replied:
Ok. we can use `oneWay` instead. https://www.emberjs.com/api/ember/3.0/functions/@ember%2Fobject%2Fcomputed/oneWay

Xuefeng-Zhu replied:
updated

balajimanoharan commented on packages/visualizations/addon/components/number-format-dropdown.js:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * @action onUpdateFormat
+     */
+    onUpdateFormat({ format }) {
```
If the `number-format-selector` is only going to update the format, do we need to wrap `format` in an object?

Xuefeng-Zhu replied:
object is required for metric label config

Xuefeng-Zhu replied:
`{{number-format-selector format=options.format onUpdateFormat=onUpdateConfig}}` metric label is passing in `onUpdateConfig` directly. If we want to wrap it up, we need another action in metric label. 

IMHO, returning object is more flexible than format string

balajimanoharan replied:
If there is a need for an object then it would be alright, but I'm not sure there is a need for it. 
Can we discuss this?

Xuefeng-Zhu replied:
OK. I will make it pass string instead of object.

balajimanoharan commented on packages/visualizations/addon/components/number-format-dropdown.js:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * @action onUpdateFormat
+     */
+    onUpdateFormat(format) {
```
Now you can replace this with the mut helper
https://emberjs.com/api/ember/3.1/classes/Ember.Templates.helpers/methods/mut?anchor=mut

balajimanoharan commented on packages/visualizations/addon/components/number-format-selector.js:
```diff
@@ -46,15 +46,15 @@ export default Component.extend({
      */
     updateFormat(format) {
-      this.attrs.onUpdateFormat({ format });
+      this.attrs.onUpdateFormat(format);
```
Now you don't need these actions, you could 
`(action this.attrs.onUpdateFormat format)` in the template.

Same thing for clearFormat.

balajimanoharan commented on packages/visualizations/addon/templates/components/navi-visualizations/table.hbs:
```diff
@@ -15,6 +15,9 @@
                 focus-out=(action 'updateColumnDisplayName' column)
               }}
+              {{#if (or (eq column.type 'metric') (eq column.type 'threshold'))}}
+                {{number-format-dropdown class="table-header-cell__number-format-dropdown" column=column onUpdateReport=onUpdateReport}}
```
Can we say `(action onUpdateReport)`?
